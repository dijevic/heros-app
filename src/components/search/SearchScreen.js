import React, { useMemo } from 'react'
import { useLocation } from 'react-router';
import queryString from 'query-string'
import { UseForm } from '../../hooks/useForm';
import { HeroeCard } from '../heroe/HeroeCard';
import { getHeroeByName } from '../../selectors/GetHeroesByName';

export const SearchScreen = ({ history }) => {

    const location = useLocation()
    const { q = '' } = queryString.parse(location.search)


    const [formValues, handleInputChange] = UseForm({
        searchText: q
    })

    const { searchText } = formValues
    const heroesFiltrados = useMemo(() => getHeroeByName(q), [q])
    const handleSearch = (e) => {

        e.preventDefault();
        history.push(`?q=${searchText}`)

    }
    return (
        <div>
            <h1>Search</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <h4>search Form</h4>
                    <hr />


                    <form onSubmit={handleSearch}>
                        <input
                            className="form-control"
                            placeholder="find tour hero"
                            type="text"
                            name="searchText"
                            value={searchText}
                            onChange={handleInputChange}
                        />


                        <button className="btn btn-primary d-inline-block w-100 mt-3" type="submit">Buscar</button>
                    </form>
                </div>

                <div className="col-7 card-columns row row-cols-1 row-cols-md-3 g-4 ">
                    <h4 className="text-center w-100">results</h4>

                    {
                        (q === '') &&
                        <div className="alert alert-info w-100">
                            Busca un heroe
                        </div>
                    }
                    {
                        (q !== '' && heroesFiltrados.length === 0) &&
                        <div className="alert alert-info w-100">
                            no existe un heroe con el nombre {q}
                        </div>
                    }

                    {heroesFiltrados.map(hero => (
                        <HeroeCard key={hero.id} {...hero} />
                    ))}
                </div>
            </div>
        </div>
    )
}
