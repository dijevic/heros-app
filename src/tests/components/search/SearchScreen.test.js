import { mount } from "enzyme"
import { MemoryRouter, Route } from "react-router"
import { SearchScreen } from "../../../components/search/SearchScreen"

describe('pruebas en mi componente de Search screen', () => {

    test('debe hacer match con el snapshot por defecto', () => {
        const wrapper = mount(
            <MemoryRouter>
                <SearchScreen />
            </MemoryRouter>
        )
        expect(wrapper.find('.alert-info').text().trim()).toBe('Busca un heroe')
        expect(wrapper).toMatchSnapshot()
    })

    test('debe de leer y buscar por parametro', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path="/search">
                    <SearchScreen />
                </Route>
            </MemoryRouter>
        )
        expect(wrapper.find('input').prop('value')).toBe('batman')
        expect(wrapper).toMatchSnapshot()

    })

    test('debe mostrar un error si la ruta de busqueda o existe', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=1234']}>
                <Route path="/search">
                    <SearchScreen />
                </Route>
            </MemoryRouter>

        )
        expect(wrapper.find('.alert-info').text().trim()).toBe('no existe un heroe con el nombre 1234')
        expect(wrapper).toMatchSnapshot()
    })

    test('debe llamar ael history.push', () => {
        const history = {
            push: jest.fn()
        }
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=1234']}>
                <Route path="/search">
                    <SearchScreen history={history} />
                </Route>
            </MemoryRouter>

        )

        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman'
            }
        })

        wrapper.find('form').simulate('submit')
        expect(history.push).toHaveBeenCalledWith(`?q=batman`)


    })
})