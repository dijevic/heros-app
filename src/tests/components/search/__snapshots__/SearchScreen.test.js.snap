// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en mi componente de Search screen debe de leer y buscar por parametro 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="find tour hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-primary d-inline-block w-100 mt-3"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7 card-columns row row-cols-1 row-cols-md-3 g-4 "
    >
      <h4
        className="text-center w-100"
      >
        results
      </h4>
      <div
        className="my-card"
        onClick={[Function]}
      >
        <img
          alt="Batman"
          className="img img-responsive"
          src="./assets/heroes/dc-batman.jpg"
        />
        <div
          className="profile-name"
        >
          Batman
        </div>
        <div
          className="profile-position"
        >
          Bruce Wayne
        </div>
        <div
          className="profile-overview"
        >
          <div
            className="profile-overview"
          >
            <div
              className="row"
            >
              <div
                className="col-ms-4"
              >
                <h3>
                  DC Comics
                </h3>
                <p>
                  Primera aparici√≥n: 
                  <br />
                  Detective Comics #27
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`pruebas en mi componente de Search screen debe hacer match con el snapshot por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="find tour hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-primary d-inline-block w-100 mt-3"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7 card-columns row row-cols-1 row-cols-md-3 g-4 "
    >
      <h4
        className="text-center w-100"
      >
        results
      </h4>
      <div
        className="alert alert-info w-100"
      >
        Busca un heroe
      </div>
    </div>
  </div>
</div>
`;

exports[`pruebas en mi componente de Search screen debe mostrar un error si la ruta de busqueda o existe 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="find tour hero"
          type="text"
          value="1234"
        />
        <button
          className="btn btn-primary d-inline-block w-100 mt-3"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7 card-columns row row-cols-1 row-cols-md-3 g-4 "
    >
      <h4
        className="text-center w-100"
      >
        results
      </h4>
      <div
        className="alert alert-info w-100"
      >
        no existe un heroe con el nombre 
        1234
      </div>
    </div>
  </div>
</div>
`;
